updater/src/updater.rs

```diff
@@ -248,189 +301,406 @@ impl Updater {
+				Err(err) => {
+					let delay = 2usize.pow(retries) as u64;
+					// cap maximum backoff to 1 month
+					let delay = cmp::min(delay, 30 * 24 * 60 * 60);
+					let backoff = (retries, self.time_provider.now() + Duration::from_secs(delay));
+
+					state.status = UpdaterStatus::FetchBackoff { release: release.clone(), backoff, binary };
+
+					warn!("Unable to fetch update ({}): {:?}, retrying in {} seconds.", release.version, err, delay);
+				},
+			}
+		}
+	}
+
+	fn updater_step(&self) {
+		let current_block_number = self.client.upgrade().map_or(0, |c| c.block_number(BlockId::Latest).unwrap_or(0));
+
+		let mut state = self.state.lock();
+
+		if let Some(latest) = state.latest.clone() {
+			let fetch = |binary| {
+				info!(target: "updater", "Attempting to get parity binary {}", binary);
+				let weak_self = self.weak_self.lock().clone();
+				let latest = latest.clone();
+				let f = move |res: Result<PathBuf, fetch::Error>| {
+					if let Some(this) = weak_self.upgrade() {
+						this.on_fetch(&latest, res)
+					}
 				};
-				s.ready = Some(fetched);
-				auto
+
+				self.fetcher.fetch(
+					binary,
+					fetch::Abort::default().with_max_size(self.update_policy.max_size),
+					Box::new(f));
 			};
-			if auto {
-				// will lock self.state, so ensure it's outside of previous block.
-				self.execute_upgrade();
+
+			match state.status.clone() {
+				// updater is disabled
+				UpdaterStatus::Disabled => {},
+				// the update has already been installed
+				UpdaterStatus::Installed { ref release, .. } if *release == latest.track => {},
+				// we're currently fetching this update
+				UpdaterStatus::Fetching { ref release, .. } if *release == latest.track => {},
+				// the fetch has failed and we're backing off the next retry
+				UpdaterStatus::FetchBackoff { ref release, backoff, .. } if *release == latest.track && self.time_provider.now() < backoff.1 => {},
+				// we're delaying the update until the given block number
+				UpdaterStatus::Waiting { ref release, block_number, .. } if *release == latest.track && current_block_number < block_number => {},
+				// we're at (or past) the block that triggers the update, let's fetch the binary
+				UpdaterStatus::Waiting { ref release, block_number, binary } if *release == latest.track && current_block_number >= block_number => {
+					info!(target: "updater", "Update for binary {} triggered", binary);
+
+					state.status = UpdaterStatus::Fetching { release: release.clone(), binary, retries: 1 };
+					// will lock self.state
+					drop(state);
+					fetch(binary);
+				},
+				// we're ready to retry the fetch after we applied a backoff for the previous failure
+				UpdaterStatus::FetchBackoff { ref release, backoff, binary } if *release == latest.track && self.time_provider.now() >= backoff.1 => {
+					state.status = UpdaterStatus::Fetching { release: release.clone(), binary, retries: backoff.0 + 1 };
```

Comment:
Should we increase the backoff even if `latest.track` is now different?

Reply:
If `latest.track` changes then we don't enter this branch and instead start downloading the new release (I think there's a test for this).